### PR TITLE
Add connect timeout

### PR DESCRIPTION
**Pull request by kaikash**
https://github.com/ytsaurus/ytsaurus/issues/641

### CODE REVIEW

File: yt/yt/core/net/dialer.cpp [hunk 1]
```diff
@@ -287,10 +292,14 @@ class TAsyncDialerSession
             return;
         }
 
-        if (Config_->EnableAggressiveReconnect) {
+        if (Config_->EnableAggressiveReconnect || Deadline_) {
+            const auto deadline = Min(
```
Comment (maxim-babenko) [inline yt/yt/core/net/dialer.cpp:296]: Typically we don't use const for locals

File: yt/yt/core/net/dialer.cpp [hunk 2]
```diff
@@ -287,10 +292,14 @@ class TAsyncDialerSession
...
+                Deadline_ ? *Deadline_ : TInstant::Max(),
```
Comment (maxim-babenko) [inline yt/yt/core/net/dialer.cpp:297]: Let's use value_or

File: yt/yt/core/net/dialer.cpp [hunk 3]
```diff
@@ -359,6 +368,16 @@ class TAsyncDialerSession
             Timeout_ *= Config_->RtoScale * GetRandomVariation();
         }
 
+        if (Deadline_ && TInstant::Now() >= *Deadline_) {
+            auto error = TError(NRpc::EErrorCode::TransportError, "Connect timed out")
+                << TErrorAttribute("Timeout", Config_->ConnectTimeout);
```
Comment (maxim-babenko) [inline yt/yt/core/net/dialer.cpp:373]: Timeout -> timeout (attributes are in lower case)

Comment (maxim-babenko) [reply yt/yt/core/net/dialer.cpp:373]: "Connect timed out" above vs "Connect timeout" below -- let's choose a single spelling

**Review by maxim-babenko — CHANGES_REQUESTED** [review]

**Comment by maxim-babenko** [comment]
Shipped!

File: yt/yt/core/net/config.cpp [hunk 4]
```diff
@@ -17,6 +17,7 @@ void TDialerConfig::Register(TRegistrar registrar)
     registrar.Parameter("rto_scale", &TThis::RtoScale)
         .GreaterThan(0.0)
         .Default(2.0);
+    registrar.Parameter("connect_timeout", &TThis::ConnectTimeout);
```
Comment (maxim-babenko) [inline yt/yt/core/net/config.cpp:20]: Please add ``Default()`` here

**Review by maxim-babenko — CHANGES_REQUESTED** [review]

**Review by Kontakter — APPROVED** [review]
To run internal tests

File: yt/yt/core/net/config.cpp [hunk 5]
```diff
@@ -17,6 +17,8 @@ void TDialerConfig::Register(TRegistrar registrar)
     registrar.Parameter("rto_scale", &TThis::RtoScale)
         .GreaterThan(0.0)
         .Default(2.0);
+    registrar.Parameter("connect_timeout", &TThis::ConnectTimeout)
+        .Default();
```
[the remark this answers is not in the document]
Comment (kaikash) [reply yt/yt/core/net/config.cpp]: The reason I haven't set any timeout here is that doing so would change the current behavior of internal clusters and could potentially cause some problems.

[the remark this answers is not in the document]
Comment (gritukan) [reply yt/yt/core/net/config.cpp]: I see. Honestly I do not see any way how reasonable timeout can break something here, but let's wait for code maintainers.

**Comment by maxim-babenko** [comment]
Shipped!

File: yt/yt/core/net/config.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (maxim-babenko) [reply yt/yt/core/net/config.cpp]: 15 secs seems a reasonable limit, let's try this revision

[the remark this answers is not in the document]
Comment (gritukan) [reply yt/yt/core/net/config.cpp]: Discussed with @maxim-babenko, 15 seconds should be reasonable in all cases.

File: yt/yt/core/net/config.h [hunk 6]
```diff
@@ -22,6 +22,7 @@ class TDialerConfig
     TDuration MinRto;
     TDuration MaxRto;
     double RtoScale;
+    std::optional<TDuration> ConnectTimeout;
```
[the remark this answers is not in the document]
Comment (kaikash) [reply yt/yt/core/net/config.h]: Makes sense. Also, renamed `Timeout_` to `ReconnectTimeout_` for a better readability.

File: yt/yt/core/net/dialer.cpp [hunk 7]
```diff
@@ -355,12 +359,22 @@ class TAsyncDialerSession
...
+            YT_LOG_ERROR(error);
+            Finished_ = true;
+            guard.Release();
+            OnFinished_(error);
+            return;
         }
 
-        YT_LOG_DEBUG("Connect timeout; trying to reconnect (Timeout: %v)",
-            Timeout_);
+        YT_LOG_DEBUG("Connect timeout; trying to reconnect (Reconnect timeout: %v)",
```
Comment (gritukan) [inline yt/yt/core/net/dialer.cpp:376]: ReconnectTimeout

**Comment by maxim-babenko** [comment]
Shipped!

**Pull request closed without merging**